### PR TITLE
feat(fs): add virtual readonly mounts for session filesystem

### PR DESCRIPTION
**Pull request by chaliy**
## What
Add `MountSource::Virtual` variant and `VirtualMountRegistry` so capabilities can serve files from memory without writing rows to `session_files`. This is the core plumbing for virtual readonly mounts.

## Why
Platform chat sessions will mount docs (~66 files, ~5 MB) that are identical across all sessions — static, compiled into the binary. Writing 66 INSERT per session creation is unnecessary. Virtual mounts eliminate per-session DB cost for static content. Future use cases: starter templates, SDK examples, skill definitions.

Closes EVE-261 (core plumbing — `platform_docs` consumer capability is a follow-up).

## How
**Core types (`crates/core/src/capability_types.rs`):**
- `VirtualFileTree`: path-indexed readonly tree (`HashMap`)
  - `insert_text()`, `get()`, `list_directory()`, `all_files()` (for grep)
  - Auto-creates parent directories on insert
- `MountSource::Virtual { tree: Arc }`: new enum variant

**Registry (`crates/server/src/services/virtual_mount_registry.rs`):**
- `VirtualMountRegistry`: per-session registry using `parking_lot::RwLock`
- `register()`, `evict()`, `read_file()`, `list_directory()`, `is_virtual_path()`, `grep()`
- `VirtualFileMeta`: metadata-only struct for directory listings (avoids cloning content)

**Session file service integration:**
- `read_file()`: virtual-first lookup, fall back to DB
- `stat()`: virtual-first lookup, fall back to DB
- `list_directory()`: merge virtual + DB entries (virtual wins on conflict), sorted dirs-first then by path
- `update_file()`: reject writes to virtual mount paths
- `apply_single_mount()`: registers virtual mounts in-memory instead of writing to DB; rejects ReadWrite access; errors if no registry configured

## Risk
- Low
- No runtime behavior change until a capability uses `MountSource::Virtual`
- Existing inline mount path is completely unchanged
- Follow-up: `platform_docs` capability, grep integration for `DirectWorkerAdapters`/`DbSessionFileStore`

## Security
- Threat categories reviewed: TM-FS, TM-TENANT, TM-DOS
- TM-FS: Paths validated via existing `normalize_path`/`validate_path` before reaching virtual layer. Virtual files hardcoded `is_readonly: true`. Write protection added to `update_file`. `strip_mount_prefix` uses segment-boundary matching (no partial path leaks).
- TM-TENANT: Virtual mounts keyed by session_id (UUID). No cross-session access.
- TM-DOS: Trees built once at startup, shared via `Arc`. Grep iterates bounded static content. Directory listings return metadata-only (`VirtualFileMeta`), avoiding content cloning.
- No new crate dependencies (uses existing `parking_lot`, `regex`).

## Checklist
- [x] Tests added or updated (18 new tests: 7 session file service, 11 virtual mount registry)
- [x] Backward compatibility considered (additive only, no breaking changes)
- [x] Security review performed against relevant threat model categories
- [x] All review comments addressed (code change or written reasoning)

### CODE REVIEW

File: crates/server/src/services/virtual_mount_registry.rs [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+/// Strip the mount prefix from a path. Returns the relative portion.
+/// E.g. strip_mount_prefix("/docs/foo.md", "/docs") -> Some("foo.md")
+/// E.g. strip_mount_prefix("/docs", "/docs") -> Some("")
+fn strip_mount_prefix<'a>(path: &'a str, mount_path: &str) -> Option<&'a str> {
+    if path == mount_path {
+        Some("")
+    } else if let Some(rest) = path.strip_prefix(mount_path) {
+        rest.strip_prefix('/').or(Some(rest))
+    } else {
+        None
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/server/src/services/virtual_mount_registry.rs]: Fixed. Applied the suggested segment-boundary approach: exact match, root special-case, or `{mount_path}/` prefix. Added `strip_mount_prefix_no_match_partial_segment` and `registry_read_file_no_partial_segment_match` tests to verify `/docs` no longer matches `/docs2`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/server/src/services/virtual_mount_registry.rs [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        };
+        for mount in session_mounts {
+            if strip_mount_prefix(dir_path, &mount.mount_path).is_some() {
+                for (entry_path, file) in mount.tree.list_directory(dir_path) {
+                    results.push(VirtualFileRead {
+                        content: file.content.clone(),
+                        is_directory: file.is_directory,
+                        path: entry_path,
+                    });
+                }
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/server/src/services/virtual_mount_registry.rs:95]: Fixed. Introduced `VirtualFileMeta` struct with only `path`, `is_directory`, `size_bytes`. `list_directory()` now returns `Vec<VirtualFileMeta>` instead of cloning content. Caller in `SessionFileService::list_directory` uses `vf.size_bytes` directly.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/server/src/services/session_file.rs [hunk 3]
```diff
@@ -262,6 +274,28 @@ impl SessionFileService {
...
+                path: vf.path.clone(),
+                name: FileInfo::name_from_path(&vf.path),
+                content: Some(content),
+                encoding,
+                is_directory: vf.is_directory,
+                is_readonly: true,
+                size_bytes: vf.content.len() as i64,
+                created_at: now,
+                updated_at: now,
+            }));
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/server/src/services/session_file.rs:301]: Fixed. Virtual directory reads now return `content: None` (skipping `encode_content`) when `vf.is_directory` is true, matching DB-backed directory behavior. Added `virtual_read_directory_returns_none_content` test.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/server/src/services/session_file.rs [hunk 4]
```diff
@@ -312,10 +367,34 @@ impl SessionFileService {
...
+                        session_id,
+                        path: vf.path,
+                        name,
+                        is_directory: vf.is_directory,
+                        is_readonly: true,
+                        size_bytes: vf.content.len() as i64,
+                        created_at: now,
+                        updated_at: now,
+                    });
+                }
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/server/src/services/session_file.rs]: Fixed. Changed to `entries.retain(|e| e.name != name)` before pushing the virtual entry, so virtual actually wins on conflict. Added `virtual_wins_on_name_conflict` test that creates a DB file and virtual file at the same path, then verifies the virtual (readonly) entry is returned.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/server/src/services/session_file.rs [hunk 5]
```diff
@@ -312,10 +367,34 @@ impl SessionFileService {
...
+            }
+        }
+
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/server/src/services/session_file.rs:401]: Fixed. Added the suggested sort after merging: `entries.sort_by(|a, b| b.is_directory.cmp(&a.is_directory).then_with(|| a.path.cmp(&b.path)))`. Added `virtual_list_directory_sorted_dirs_first` test.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/server/src/services/session_file.rs [hunk 6]
```diff
@@ -650,6 +736,15 @@ impl SessionFileService {
         let is_readonly = mount.is_readonly();
         let mut stats = MountStats::default();
 
+        // Virtual mounts are registered in-memory, not written to DB
+        if let MountSource::Virtual { tree } = &mount.source {
+            if let Some(registry) = &self.virtual_registry {
+                registry.register(session_id, path, tree.clone(), mount.capability_id.clone());
+                stats.files_created += tree.len();
+            }
+            return Ok(stats);
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/server/src/services/session_file.rs:769]: Fixed. Now returns `Err(anyhow!("virtual mount at '...' for capability '...' cannot be applied: no virtual registry configured"))` when `virtual_registry` is `None`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/core/src/capability_types.rs [hunk 7]
```diff
@@ -156,6 +157,125 @@ impl From<String> for AgentCapabilityConfig {
...
+
+impl Default for VirtualFileTree {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl PartialEq for VirtualFileTree {
+    fn eq(&self, other: &Self) -> bool {
+        self.files.len() == other.files.len()
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/core/src/capability_types.rs]: Fixed. Derived `PartialEq, Eq` on `VirtualFile` and changed `VirtualFileTree::eq` to use `self.files == other.files` for full structural equality.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/server/src/services/session_file.rs [hunk 8]
```diff
@@ -336,6 +415,13 @@ impl SessionFileService {
     ) -> Result<Option<SessionFile>> {
         let path = Self::normalize_path(path);
 
+        // Virtual files cannot be modified
+        if let Some(registry) = &self.virtual_registry
+            && registry.is_virtual_path(&session_id, &path)
+        {
+            return Err(anyhow!("Cannot modify readonly file: {}", path));
+        }
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/server/src/services/session_file.rs:435]: Fixed. Added 7 focused tests in `session_file::tests`: `virtual_read_file_returns_content`, `virtual_read_directory_returns_none_content`, `virtual_stat_returns_metadata`, `virtual_list_directory_returns_entries`, `virtual_list_directory_sorted_dirs_first`, `virtual_wins_on_name_conflict`, `virtual_update_file_rejected`. Also added 11 unit tests in `virtual_mount_registry::tests` covering `strip_mount_prefix`, read, list, evict, grep, and segment-boundary edge cases.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/server/src/services/session_file.rs [hunk 9]
```diff
@@ -685,6 +780,10 @@ impl SessionFileService {
                             .await?;
                     }
                 }
+                MountSource::Virtual { .. } => {
+                    // Virtual mounts are handled in apply_single_mount, not here
+                    unreachable!("Virtual mounts should be handled before apply_mount_source");
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/server/src/services/session_file.rs]: Fixed. Replaced `unreachable!` with `return Err(anyhow!("Virtual mounts are only supported at the mount root"))`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: crates/server/src/services/session_file.rs [hunk 10]
```diff
@@ -650,6 +736,15 @@ impl SessionFileService {
         let is_readonly = mount.is_readonly();
         let mut stats = MountStats::default();
 
+        // Virtual mounts are registered in-memory, not written to DB
+        if let MountSource::Virtual { tree } = &mount.source {
+            if let Some(registry) = &self.virtual_registry {
+                registry.register(session_id, path, tree.clone(), mount.capability_id.clone());
+                stats.files_created += tree.len();
+            }
```
[the remark this answers is not in the document]
Comment (chaliy) [reply crates/server/src/services/session_file.rs:768]: Fixed. Added an early check in `apply_single_mount`: if `mount.access == MountAccess::ReadWrite` for a virtual mount, returns `Err(anyhow!("Virtual mounts are always read-only; mount at '...' has ReadWrite access"))`.

---
_Generated by [Claude Code](https://claude.ai/code)_
